akka-http-core/src/main/scala/akka/http/scaladsl/OutgoingConnectionBuilder.scala

```diff
@@ -0,0 +1,110 @@
+   */
+  def toPort(port: Int): OutgoingConnectionBuilder
+
+  def enableHttps(context: HttpsConnectionContext): OutgoingConnectionBuilder
+
+  /**
+   * Use a specific local host and port to create the outgoing connection from
+   */
+  def withLocalAddress(localAddress: InetAddress): OutgoingConnectionBuilder
+
+  /**
+   * Use a custom [[ConnectionContext]] for the connection.
+   */
+  def withConnectionContext(context: ConnectionContext): OutgoingConnectionBuilder
+
+  /**
+   * Use custom [[ClientConnectionSettings]] for the connection.
+   */
+  def withClientConnectionSettings(settings: ClientConnectionSettings): OutgoingConnectionBuilder
+
+  /**
+   * Use a custom logger
+   */
+  def logTo(logger: LoggingAdapter): OutgoingConnectionBuilder
+
+  /**
+   * Create flow that when materialized creates a single connection to the HTTP server.
+   *
+   * Note that the responses are not guaranteed to arrive in the same order as the requests go out (In the case of a HTTP/2 server)
+   * so therefore requests needs to have a [[akka.http.scaladsl.model.http2.RequestResponseAssociation]]
+   * which Akka HTTP will carry over to the corresponding response for a request.
+   */
+  def connectionFlow(): Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]]
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object OutgoingConnectionBuilder {
+
+  def apply(host: String, port: Int, system: ClassicActorSystemProvider): OutgoingConnectionBuilder =
+    Impl(
+      host,
+      port,
+      clientConnectionSettings = ClientConnectionSettings(system),
+      connectionContext = ConnectionContext.noEncryption(),
+      log = system.classicSystem.log,
+      localAddress = None
+    )
+
+  private case class Impl(
+    host:                     String,
+    port:                     Int,
+    clientConnectionSettings: ClientConnectionSettings,
+    connectionContext:        ConnectionContext,
+    log:                      LoggingAdapter,
+    localAddress:             Option[InetAddress]) extends OutgoingConnectionBuilder {
+
+    override def toHost(host: String): OutgoingConnectionBuilder = copy(host = host)
+
+    override def toPort(port: Int): OutgoingConnectionBuilder = copy(port = port)
+
+    override def enableHttps(context: HttpsConnectionContext): OutgoingConnectionBuilder = copy(connectionContext = context)
+
+    override def withLocalAddress(localAddress: InetAddress): OutgoingConnectionBuilder = copy(localAddress = Some(localAddress))
+
+    override def withConnectionContext(context: ConnectionContext): OutgoingConnectionBuilder = copy(connectionContext = context)
+
+    override def withClientConnectionSettings(settings: ClientConnectionSettings): OutgoingConnectionBuilder = copy(clientConnectionSettings = settings)
+
+    override def logTo(logger: LoggingAdapter): OutgoingConnectionBuilder = copy(log = logger)
+
+    override def connectionFlow(): Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]] = ???
```

Comment:
Some questions:
 * How would the actual protocol negotiation work?
 * Should the user be able to explicitly choose between HTTP/1.1 and HTTP/2 as well?

Reply:
I like the builder approach, indeed consistent with the new server API's, and gives the opportunity to introduce additional options later. Might get tricky when some options are mutually exclusive, I guess, but OK.

I definitely think it's useful to be able to explicitly choose between various connection/negotiation types, but maybe just start with one?

I agree it's not super obvious the `Flow[HttpRequest, HttpResponse]` is *not* ordered. Perhaps rather than renaming `connectTo` we could rename `connectionFlow` to something that conveys it's unordered? Not sure of a great name though.

An HTTP2 connection can have a max number of concurrent streams (https://tools.ietf.org/html/rfc7540#section-6.5.2). Since each `HttpResponse` corresponds to a stream, you could in theory get stuck if you're not consuming the HttpResponses. Would that cause trouble if you're for example trying to populate a tree-like structure with a depth-first streaming approach? Or nudge you towards a better approach anyway? Not sure, but this seems like a good start :)

Reply:
One more question, akka-http2-support depends on akka-http-core where Http lives, so it is not possible to just add a dependency to get access to the HTTP/2 classes. Is the long term goal to keep them as separate modules or to merge them or how do we make using HTTP/2 more seamless moving forward?

Reply:
Moved it to the http2-support module for now

Reply:
For those reading along: for the server-side Http2 support, the actual implementation is in `Http2Ext` in the `akka-http2-support` module, available via `Http2()`.

We use a reflection trick to use that implementation from the regular `Http` extension when the module is available and enabled. We could possibly do the same for the client API. For now having it just in the http2-support module might be the quickest way to iterate, though.

Reply:
See also #3514